src/createSelector.js

```diff
@@ -44,69 +43,69 @@ const getIdSelector = dependencies => {
     : getCombinedIdSelector(uniqIdSelectors)
 }
 
-const getComputeFn = (dependencies_, computeFn, idSelector, getCache) => {
+const getComputeFn = (
+  dependencies_,
+  computeFn,
+  idSelector,
+  equalityFn,
+  globalCache
+) => {
   const dependencies = dependencies_.length > 0 ? dependencies_ : ofIdentity
   let nComputations = 0
 
-  if (!getCache) {
-    const cache = new Array(2)
-    getCache = () => cache
-  }
+  const getCache = !idSelector
+    ? () => globalCache
+    : (...args) => {
+        const id = idSelector(...args)
+        return globalCache[id] || (globalCache[id] = new Array(2))
+      }
 
-  const computeFnCached = (computeFnArgs, id) => {
-    const cache = getCache(id)
+  const resFn = (...args) => {
+    const cache = getCache(...args)
     const [prevArgs, prevRes] = cache
-    if (prevArgs && computeFnArgs.every((val, idx) => val === prevArgs[idx])) {
+    const computedArgs = dependencies.map(fn => fn(...args))
+    if (prevArgs && computedArgs.every((val, idx) => val === prevArgs[idx])) {
       return prevRes
     }
+    cache[0] = computedArgs
+    const res = computeFn(...computedArgs)
     nComputations++
-    const res = computeFn(...computeFnArgs)
-    cache[0] = computeFnArgs
-    cache[1] = res
-    return res
+    return (cache[1] = equalityFn && equalityFn(res, prevRes) ? prevRes : res)
   }
 
-  const resFn = idSelector
-    ? (...args) =>
-        computeFnCached(
-          dependencies.map(fn => fn(...args)),
-          idSelector(...args)
-        )
-    : (...args) => computeFnCached(dependencies.map(fn => fn(...args)))
-
   resFn.recomputations = () => nComputations
   resFn.resetRecomputations = () => (nComputations = 0)
   resFn.dependencies = dependencies
   resFn.resultFunc = computeFn
-  resFn.resultFuncCached = computeFnCached
-  resFn.resultFuncCached.recomputations = resFn.recomputations
   return resFn
 }
 
 const getInstanceSelector = (
   dependencies,
   computeFn,
   idSelector,
-  cache = new Map()
+  equalityFn
 ) => {
-  const usages = new Map()
-
-  const result = getComputeFn(dependencies, computeFn, idSelector, id => {
-    if (!cache.has(id)) cache.set(id, new Array(2))
-    return cache.get(id)
-  })
+  const cache = {}
+  let usages = {}
+
+  const result = getComputeFn(
+    dependencies,
+    computeFn,
+    idSelector,
+    equalityFn,
+    cache
+  )
 
   result.idSelector = idSelector
```

Comment:
Because if I set it inside `getComputeFn` then the resulting selector of a non-instance selector will have a property that shouldn't be there... Yep, it's value will be `undefined`, but the property would be there :sweat_smile:.

Also, with the current implementation `getComputeFn` is only responsible for setting the properties that are common among all selectors, while `getInstanceSelector` is responsible for setting those properties that are specific to the "instance selectors" and this is one of the properties that it's specific to "instance selectors".